File: server/routes/bids.js
import { v4 as uuidv4 } from "uuid";
import { consignments, incrementBidCount } from "./consignments.js";
import Bid from '../models/Bid.js';
import Consignment from '../models/Consignment.js';

// In-memory database (fallback if MongoDB is not available)
const bids = new Map();

export const handleCreateBid = async (req, res) => {
  try {
    console.log("[CreateBid] user:", req.user);
    console.log("[CreateBid] body:", req.body);
    
    const { consignmentId, bidAmount, estimatedDelivery, notes } = req.body;

    // Validation
    if (!consignmentId || !bidAmount || !estimatedDelivery) {
      return res.status(400).json({
        success: false,
        message:
          "Consignment ID, bid amount, and estimated delivery are required",
      });
    }

    if (req.user.userType !== "msme") {
      return res.status(403).json({
        success: false,
        message: "Only MSMEs can place bids",
      });
    }

    // Try MongoDB first, fallback to in-memory storage
    try {
      // Check if consignment exists and is open in MongoDB
      const consignment = await Consignment.findById(consignmentId);
      if (!consignment) {
        return res.status(404).json({
          success: false,
          message: "Consignment not found",
        });
      }

      if (consignment.status !== "open") {
        return res.status(400).json({
          success: false,
          message: "This consignment is no longer accepting bids",
        });
      }

      // Check if user already bid on this consignment
      const existingBid = await Bid.findOne({
        consignmentId: consignmentId,
        bidderId: req.user._id
      });

      if (existingBid) {
        return res.status(400).json({
          success: false,
          message: "You have already placed a bid on this consignment",
        });
      }

      // Validate bid amount
      if (parseFloat(bidAmount) > consignment.budget) {
        return res.status(400).json({
          success: false,
          message: "Bid amount cannot exceed the maximum budget",
        });
      }

      // Validate estimated delivery
      const estimatedDate = new Date(estimatedDelivery);
      const deadline = new Date(consignment.deadline);
      if (estimatedDate > deadline) {
        return res.status(400).json({
          success: false,
          message: "Estimated delivery cannot be after the deadline",
        });
      }

      // Create bid in MongoDB
      const bid = new Bid({
        consignmentId,
        bidderId: req.user._id,
        bidderName: req.user.name,
        bidderCompany: req.user.companyName || req.user.name,
        bidderEmail: req.user.email,
        amount: parseFloat(bidAmount),
        estimatedDeliveryTime: Math.ceil((estimatedDate - new Date()) / (1000 * 60 * 60 * 24)), // days
        status: "pending",
        message: notes || "",
      });

      await bid.save();

      // Increment bid count on consignment
      await incrementBidCount(consignmentId);

      console.log("[CreateBid] Bid created in MongoDB:", bid);

      res.status(201).json({
        success: true,
        bid,
        message: "Bid placed successfully",
      });
    } catch (mongoError) {
      console.log("[CreateBid] MongoDB failed, using in-memory storage:", mongoError.message);
      
      // Fallback to in-memory storage
      const consignment = consignments.get(consignmentId);
      if (!consignment) {
        return res.status(404).json({
          success: false,
          message: "Consignment not found",
        });
      }

      if (consignment.status !== "open") {
        return res.status(400).json({
          success: false,
          message: "This consignment is no longer accepting bids",
        });
      }

      // Check if user already bid on this consignment
      const existingBid = Array.from(bids.values()).find(
        (bid) =>
          bid.consignmentId === consignmentId && bid.bidderId === req.user.id,
      );

      if (existingBid) {
        return res.status(400).json({
          success: false,
          message: "You have already placed a bid on this consignment",
        });
      }

      // Validate bid amount
      if (parseFloat(bidAmount) > consignment.budget) {
        return res.status(400).json({
          success: false,
          message: "Bid amount cannot exceed the maximum budget",
        });
      }

      // Validate estimated delivery
      const estimatedDate = new Date(estimatedDelivery);
      const deadline = new Date(consignment.deadline);
      if (estimatedDate > deadline) {
        return res.status(400).json({
          success: false,
          message: "Estimated delivery cannot be after the deadline",
        });
      }

      // Create bid in memory
      const bidId = uuidv4();
      const bid = {
        id: bidId,
        consignmentId,
        consignmentTitle: consignment.title,
        bidderId: req.user.id,
        bidderName: req.user.name,
        bidderCompany: req.user.companyName || req.user.name,
        bidAmount: parseFloat(bidAmount),
        estimatedDelivery,
        notes: notes || "",
        status: "pending",
        createdAt: new Date().toISOString(),
      };

      bids.set(bidId, bid);

      // Increment bid count on consignment
      incrementBidCount(consignmentId);

      console.log("[CreateBid] Bid created in memory:", bid);

      res.status(201).json({
        success: true,
        bid,
        message: "Bid placed successfully (in-memory storage)",
      });
    }
  } catch (error) {
    console.error("Create bid error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleGetMyBids = async (req, res) => {
  try {
    if (req.user.userType !== "msme") {
      return res.status(403).json({
        success: false,
        message: "Only MSMEs can view their bids",
      });
    }

    try {
      const userBids = await Bid.find({ bidderId: req.user._id })
        .populate('consignmentId', 'title origin destination')
        .sort({ createdAt: -1 });

      res.json({
        success: true,
        bids: userBids,
      });
    } catch (mongoError) {
      console.log("[GetMyBids] MongoDB failed, using in-memory storage");
      
      // Fallback to in-memory storage
      const userBids = Array.from(bids.values()).filter(
        (bid) => bid.bidderId === req.user.id,
      );

      res.json({
        success: true,
        bids: userBids,
      });
    }
  } catch (error) {
    console.error("Get bids error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleGetConsignmentBids = async (req, res) => {
  try {
    const { consignmentId } = req.params;

    if (req.user.userType !== "company") {
      return res.status(403).json({
        success: false,
        message: "Only companies can view bids for their consignments",
      });
    }

    try {
      // Check if consignment belongs to the user in MongoDB
      const consignment = await Consignment.findById(consignmentId);
      if (!consignment) {
        return res.status(404).json({
          success: false,
          message: "Consignment not found",
        });
      }

      if (consignment.companyId.toString() !== req.user._id.toString()) {
        return res.status(403).json({
          success: false,
          message: "Access denied",
        });
      }

      // Get all bids for this consignment
      const consignmentBids = await Bid.find({ consignmentId })
        .populate('bidderId', 'name companyName email')
        .sort({ createdAt: -1 });

      res.json({
        success: true,
        bids: consignmentBids,
      });
    } catch (mongoError) {
      console.log("[GetConsignmentBids] MongoDB failed, using in-memory storage");
      
      // Fallback to in-memory storage
      const consignment = consignments.get(consignmentId);
      if (!consignment) {
        return res.status(404).json({
          success: false,
          message: "Consignment not found",
        });
      }

      if (consignment.companyId !== req.user.id) {
        return res.status(403).json({
          success: false,
          message: "Access denied",
        });
      }

      // Get all bids for this consignment
      const consignmentBids = Array.from(bids.values()).filter(
        (bid) => bid.consignmentId === consignmentId,
      );

      res.json({
        success: true,
        bids: consignmentBids,
      });
    }
  } catch (error) {
    console.error("Get consignment bids error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleAwardBid = async (req, res) => {
  try {
    const { bidId } = req.params;

    if (req.user.userType !== "company") {
      return res.status(403).json({
        success: false,
        message: "Only companies can award bids",
      });
    }

    try {
      // Find bid in MongoDB
      const bid = await Bid.findById(bidId);
      if (!bid) {
        return res.status(404).json({
          success: false,
          message: "Bid not found",
        });
      }

      // Check if consignment belongs to the user
      const consignment = await Consignment.findById(bid.consignmentId);
      if (consignment.companyId.toString() !== req.user._id.toString()) {
        return res.status(403).json({
          success: false,
          message: "Access denied",
        });
      }

      // Check if consignment is still open
      if (consignment.status !== "open") {
        return res.status(400).json({
          success: false,
          message: "This consignment is no longer accepting awards",
        });
      }

      // Award the bid
      bid.status = "awarded";
      bid.awardedAt = new Date();
      await bid.save();

      // Update consignment status
      consignment.status = "awarded";
      consignment.awardedBidId = bid._id;
      consignment.awardedTo = bid.bidderId;
      consignment.awardedAmount = bid.amount;
      consignment.awardedAt = new Date();
      await consignment.save();

      // Reject all other bids for this consignment
      await Bid.updateMany(
        { consignmentId: bid.consignmentId, _id: { $ne: bidId } },
        { status: "rejected" }
      );

      res.json({
        success: true,
        bid,
        message: "Bid awarded successfully",
      });
    } catch (mongoError) {
      console.log("[AwardBid] MongoDB failed, using in-memory storage:", mongoError.message);
      
      // Fallback to in-memory storage
      const bid = bids.get(bidId);
      if (!bid) {
        return res.status(404).json({
          success: false,
          message: "Bid not found",
        });
      }

      // Check if consignment belongs to the user
      const consignment = consignments.get(bid.consignmentId);
      if (consignment.companyId !== req.user.id) {
        return res.status(403).json({
          success: false,
          message: "Access denied",
        });
      }

      // Check if consignment is still open
      if (consignment.status !== "open") {
        return res.status(400).json({
          success: false,
          message: "This consignment is no longer accepting awards",
        });
      }

      // Award the bid
      bid.status = "awarded";
      bid.awardedAt = new Date().toISOString();

      // Update consignment status
      consignment.status = "awarded";
      consignment.awardedBidderId = bid.bidderId;
      consignment.finalAmount = bid.bidAmount;
      consignment.updatedAt = new Date().toISOString();

      // Reject all other bids for this consignment
      Array.from(bids.values())
        .filter((b) => b.consignmentId === bid.consignmentId && b.id !== bidId)
        .forEach((b) => {
          b.status = "rejected";
        });

      res.json({
        success: true,
        bid,
        message: "Bid awarded successfully (in-memory storage)",
      });
    }
  } catch (error) {
    console.error("Award bid error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleUpdateBidStatus = (req, res) => {
  try {
    const { bidId } = req.params;
    const { status } = req.body;

    const bid = bids.get(bidId);
    if (!bid) {
      return res.status(404).json({
        success: false,
        message: "Bid not found",
      });
    }

    // Check permissions
    const consignment = consignments.get(bid.consignmentId);
    if (
      req.user.userType === "company" &&
      consignment.companyId !== req.user.id
    ) {
      return res.status(403).json({
        success: false,
        message: "Access denied",
      });
    }

    if (req.user.userType === "msme" && bid.bidderId !== req.user.id) {
      return res.status(403).json({
        success: false,
        message: "Access denied",
      });
    }

    // Validate status
    const validStatuses = ["pending", "awarded", "rejected"];
    if (!validStatuses.includes(status)) {
      return res.status(400).json({
        success: false,
        message: "Invalid status",
      });
    }

    bid.status = status;
    bid.updatedAt = new Date().toISOString();

    res.json({
      success: true,
      bid,
      message: "Bid status updated successfully",
    });
  } catch (error) {
    console.error("Update bid error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

// Export bids for other modules (temporary, use database in production)
export { bids };

File: python_backend/routers/bids.py
from fastapi import APIRouter, HTTPException, status, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from pydantic import BaseModel
from typing import Optional, List, Dict
from datetime import datetime, timedelta
import uuid
import os
from bson import ObjectId
from db import get_bids_collection, get_consignments_collection, get_users_collection
from routers.auth import decode_token

router = APIRouter()
security = HTTPBearer()

# Pydantic models
class BidCreate(BaseModel):
    consignmentId: str
    bidAmount: float
    estimatedDelivery: str
    notes: Optional[str] = None

class BidUpdate(BaseModel):
    status: str

class BidResponse(BaseModel):
    id: str
    consignmentId: str
    consignmentTitle: str
    bidderId: str
    bidderName: str
    bidderCompany: str
    bidAmount: float
    estimatedDelivery: str
    notes: str
    status: str
    createdAt: str
    awardedAt: Optional[str] = None
    updatedAt: Optional[str] = None

class BidListResponse(BaseModel):
    success: bool
    bids: List[BidResponse]
    message: Optional[str] = None

class BidCreateResponse(BaseModel):
    success: bool
    bid: BidResponse
    message: str

async def get_current_user(credentials: HTTPAuthorizationCredentials = Depends(security)):
    """Get current user from token"""
    try:
        token = credentials.credentials
        payload = decode_token(token)
        user_id = payload.get("sub")
        
        users_collection = get_users_collection()
        user = await users_collection.find_one({"_id": ObjectId(user_id)})
        
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not found"
            )
        
        user["id"] = str(user["_id"])
        return user
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials"
        )

async def increment_bid_count(consignment_id: str):
    """Increment bid count for a consignment"""
    consignments_collection = get_consignments_collection()
    consignment = await consignments_collection.find_one({"_id": ObjectId(consignment_id)})
    if consignment:
        old_count = consignment.get("bidCount", 0)
        new_count = old_count + 1
        await consignments_collection.update_one(
            {"_id": ObjectId(consignment_id)},
            {
                "$set": {
                    "bidCount": new_count,
                    "updatedAt": datetime.now().isoformat()
                }
            }
        )
        print(f"[incrementBidCount] Updated bid count from {old_count} to {new_count}")

@router.post("/create", response_model=BidCreateResponse)
async def create_bid(
    bid_data: BidCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new bid"""
    print(f"[CreateBid] User: {current_user}")
    print(f"[CreateBid] Bid data: {bid_data}")
    
    if current_user["userType"] != "msme":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only MSMEs can place bids"
        )
    
    consignments_collection = get_consignments_collection()
    bids_collection = get_bids_collection()
    
    # Check if consignment exists and is open
    consignment = await consignments_collection.find_one({"_id": ObjectId(bid_data.consignmentId)})
    if not consignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Consignment not found"
        )
    if consignment["status"] != "open":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="This consignment is no longer accepting bids"
        )
    
    # Check if user already bid on this consignment
    existing_bid = await bids_collection.find_one({
        "consignmentId": bid_data.consignmentId,
        "bidderId": current_user["id"]
    })
    
    if existing_bid:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="You have already placed a bid on this consignment"
        )
    
    # Validate bid amount
    if bid_data.bidAmount > consignment["budget"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Bid amount cannot exceed the maximum budget"
        )
    
    # Validate estimated delivery
    try:
        estimated_date = datetime.fromisoformat(bid_data.estimatedDelivery.replace('Z', '+00:00'))
        deadline = datetime.fromisoformat(consignment["deadline"].replace('Z', '+00:00'))
        if estimated_date > deadline:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Estimated delivery cannot be after the deadline"
            )
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid date format"
        )
    
    # Create bid
    bid = {
        "consignmentId": bid_data.consignmentId,
        "consignmentTitle": consignment["title"],
        "bidderId": current_user["id"],
        "bidderName": current_user["name"],
        "bidderCompany": current_user.get("companyName") or current_user["name"],
        "bidAmount": bid_data.bidAmount,
        "estimatedDelivery": bid_data.estimatedDelivery,
        "notes": bid_data.notes or "",
        "status": "pending",
        "createdAt": datetime.now().isoformat()
    }
    
    result = await bids_collection.insert_one(bid)
    bid["id"] = str(result.inserted_id)
    
    # Increment bid count on consignment
    await increment_bid_count(bid_data.consignmentId)
    
    print(f"[CreateBid] Bid created: {bid}")
    print(f"[CreateBid] Updated consignment bid count")
    print(f"[CreateBid] Total bids in database")
    
    return BidCreateResponse(
        success=True,
        bid=BidResponse(**bid),
        message="Bid placed successfully"
    )

@router.get("/my-bids", response_model=BidListResponse)
async def get_my_bids(current_user: dict = Depends(get_current_user)):
    """Get bids placed by the current user"""
    if current_user["userType"] != "msme":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only MSMEs can view their bids"
        )
    
    bids_collection = get_bids_collection()
    
    # Get user's bids
    cursor = bids_collection.find({"bidderId": current_user["id"]})
    bids = await cursor.to_list(length=None)
    
    # Convert MongoDB documents to response format
    user_bids = []
    for bid in bids:
        bid_data = {k: v for k, v in bid.items() if k != "_id"}
        bid_data["id"] = str(bid["_id"])
        user_bids.append(BidResponse(**bid_data))
    
    return BidListResponse(
        success=True,
        bids=user_bids
    )

@router.get("/consignment/{consignment_id}", response_model=BidListResponse)
async def get_consignment_bids(
    consignment_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Get all bids for a specific consignment"""
    print(f"[GetConsignmentBids] User ID: {current_user.get('id')}")
    print(f"[GetConsignmentBids] User Type: {current_user.get('userType')}")
    print(f"[GetConsignmentBids] Consignment ID: {consignment_id}")
    consignments_db = get_consignments_db()
    bids_db = get_bids_db()

    # Check if consignment exists and belongs to the user
    if consignment_id not in consignments_db:
        print(f"[GetConsignmentBids] Consignment not found: {consignment_id}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Consignment not found"
        )

    consignment = consignments_db[consignment_id]
    print(f"[GetConsignmentBids] Consignment companyId: {consignment.get('companyId')}")

    if current_user["userType"] != "company":
        print(f"[GetConsignmentBids] Forbidden: user is not a company")
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only companies can view bids for their consignments"
        )

    if consignment["companyId"] != current_user["id"]:
        print(f"[GetConsignmentBids] Forbidden: consignment does not belong to user")
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )

    # Get all bids for this consignment
    consignment_bids = [
        BidResponse(**bid)
        for bid in bids_db.values()
        if bid["consignmentId"] == consignment_id
    ]

    print(f"[GetConsignmentBids] Found bids for consignment: {len(consignment_bids)}")
    print(f"[GetConsignmentBids] Bids: {consignment_bids}")

    return BidListResponse(
        success=True,
        bids=consignment_bids
    )

@router.post("/{bid_id}/award", response_model=BidCreateResponse)
async def award_bid(
    bid_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Award a bid (company only)"""
    if current_user["userType"] != "company":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only companies can award bids"
        )
    
    bids_db = get_bids_db()
    consignments_db = get_consignments_db()
    
    # Find bid
    if bid_id not in bids_db:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bid not found"
        )
    
    bid = bids_db[bid_id]
    
    # Check if consignment belongs to the user
    if bid["consignmentId"] not in consignments_db:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Consignment not found"
        )
    
    consignment = consignments_db[bid["consignmentId"]]
    if consignment["companyId"] != current_user["id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )
    
    # Check if consignment is still open
    if consignment["status"] != "open":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="This consignment is no longer accepting awards"
        )
    
    # Award the bid
    bid["status"] = "awarded"
    bid["awardedAt"] = datetime.now().isoformat()
    bid["updatedAt"] = datetime.now().isoformat()
    
    # Update consignment status
    consignment["status"] = "awarded"
    consignment["awardedBidderId"] = bid["bidderId"]
    consignment["finalAmount"] = bid["bidAmount"]
    consignment["updatedAt"] = datetime.now().isoformat()
    
    # Reject all other bids for this consignment
    for other_bid in bids_db.values():
        if (other_bid["consignmentId"] == bid["consignmentId"] and 
            other_bid["id"] != bid_id):
            other_bid["status"] = "rejected"
            other_bid["updatedAt"] = datetime.now().isoformat()
    
    return BidCreateResponse(
        success=True,
        bid=BidResponse(**bid),
        message="Bid awarded successfully"
    )

@router.put("/{bid_id}/status", response_model=BidCreateResponse)
async def update_bid_status(
    bid_id: str,
    status_update: BidUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Update bid status"""
    bids_db = get_bids_db()
    consignments_db = get_consignments_db()
    
    if bid_id not in bids_db:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bid not found"
        )
    
    bid = bids_db[bid_id]
    
    # Check permissions
    if bid["consignmentId"] in consignments_db:
        consignment = consignments_db[bid["consignmentId"]]
        if (current_user["userType"] == "company" and 
            consignment["companyId"] != current_user["id"]):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Access denied"
            )
    
    if (current_user["userType"] == "msme" and 
        bid["bidderId"] != current_user["id"]):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )
    
    # Validate status
    valid_statuses = ["pending", "awarded", "rejected"]
    if status_update.status not in valid_statuses:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid status"
        )
    
    bid["status"] = status_update.status
    bid["updatedAt"] = datetime.now().isoformat()
    
    return BidCreateResponse(
        success=True,
        bid=BidResponse(**bid),
        message="Bid status updated successfully"
    ) 